### ISSUE TITLE
Questions about training Voxelpose on custom dataset

### ISSUE THREAD

**Issue by zbzbzbb95** [open-mmlab/mmpose#1567]
您好，非常感谢与我们分享mmpose，在使用过程中遇到了一些问题，想请教一下。

### 我的情况：

1. 我想将voxelpose方法迁移到自制数据集多视角手部3d姿态估计任务，config文件基本与[voxelpose\panoptic\voxelpose_prn64x64x64_cpn80x80x20_panoptic_cam5.py](https://github.com/open-mmlab/mmpose/blob/master/configs/body/3d_kpt_mview_rgb_img/voxelpose/panoptic/voxelpose_prn64x64x64_cpn80x80x20_panoptic_cam5.py)一致
2. 没有预训练模型，只能从头开始训练（包括2d特征的backbone）

### 我的问题：

1. **关于freeze-2d**：

- 若想更新backbone参数就要在pipline中用GenerateHeatmapTarget与backbone（AssociativeEmbedding）输出计算loss_2d，False则会在model中冻结backbone，只用loss_pose更新3d卷积参数。
- 因为整个过程包括F.grid_sample将2d-feature投影成3d-feature都是可导的，不冻结backbone只用loss_pose在梯度更新时把2d、3d卷积的参数都更新是否具有可行性?

2. **关于nan**：

- freeze_2d设置为false，lr&warm up没有修改，batch_size 8 ，在训练约3~5个iter后，在上一个iter中pose_loss和loss_2d都很正常的情况下，下一个iter中backbone在输入正常的情况下输出heatmap为nan（使用torch.mean简单观测），进而导致loss_2d也是nan，但此时pose_loss依然有数值，总loss也为nan，依然可以训练。
- 于是我对出现nan的上一个iter进行排查，backbone输出的heatmap值很小，仔细查了一下resnet输出均值还在e+0量级，但DeconvHead输出后，均值就只有e-6量级，可是这一部分我没有任何改动，都是用的原装代码，请问您在开发过程中有试过voxelpose各阶段的输出值大致范围统计，我的nan是否和这一步数值过小有关？以及后续解决有什么建议？
- 另外：计算loss_2d 时发现了一个小bug： [在multiview_pose.py](https://github.com/open-mmlab/mmpose/blob/master/mmpose/models/detectors/multiview_pose.py) L354  `losses_2d_ = self.backbone.get_loss()`，但AssociativeEmbedding及其父类并没有get_loss函数，此处应改为`losses_2d_ = self.backbone. keypoint_head.get_loss()`,改完后运行时还有一些据类型的问题也需要相应的调整，比如list(Tensor)和Tensor的问题。

     **期待进一步回复与交流：）**

**Comment by liqikai9** [comment]
1. 关于freeze-2d：
是的，整个网络是可导的，如果 `freeze_2d = False`, 梯度依然可以反传到 2D 部分的网络，所以理论上是可行的。

2. 关于 nan

这个现象我个人猜测会不会是，同时训练 2D 和 3D 的网络比较困难。可以尝试第一阶段先用每个视角的图像预训练 2D 部分的网络，第二阶段像 VoxelPose 一样，用多视角的图像来训练 3D 部分的网络，冻结 2D 部分的网络，或者将 2D 部分网络的学习率调低一个数量级。

你阅读得很细致，这里提到的 bug 在[这个重构的 PR](https://github.com/open-mmlab/mmpose/pull/1451) 里面会修复，即将会合入 master 分支，感兴趣的话可以去试用一下～

**Comment by zbzbzbb95** [comment]
谢谢回答
关于nan的问题：
我在OptimizerHook中loss反向传播前之前增加了`torch.autograd.set_detect_anomaly(True) with torch.autograd.detect_anomaly():
runner.outputs['loss'].backward()`定位到问题在 [multiview_pose.py](https://github.com/open-mmlab/mmpose/blob/master/mmpose/models/detectors/multiview_pose.py) L115 F.grid_sample函数backward时出现nan，现在将 F.grid_sample中增加mode=‘nearest’代替默认值（可能是bilinear），还没有训练完成但已经正常训练4000iter，**在新版本中也许可以考虑借鉴**
最后，希望mmpose越来越好！

**Comment by wusize** [comment]
The bug in loss_2d has been fixed in this [pr](https://github.com/open-mmlab/mmpose/pull/1451) where the 2d part in multiview_pose was reformulated, I believe it will be merged soon.

**Comment by Taylorminer** [comment]
> 谢谢回答 关于nan的问题： 我在OptimizerHook中loss反向传播前之前增加了`torch.autograd.set_detect_anomaly(True) with torch.autograd.detect_anomaly(): runner.outputs['loss'].backward()`定位到问题在 [multiview_pose.py](https://github.com/open-mmlab/mmpose/blob/master/mmpose/models/detectors/multiview_pose.py) L115 F.grid_sample函数backward时出现nan，现在将 F.grid_sample中增加mode=‘nearest’代替默认值（可能是bilinear），还没有训练完成但已经正常训练4000iter，**在新版本中也许可以考虑借鉴** 最后，希望mmpose越来越好！

@zbzbzbb95 您好,请问将voxelpose在新场景中测试，您使用的标定的方法是什么？是怎样进行相机标定的？您使用的相机个数也是5个么？

**Comment by zbzbzbb95** [comment]
> 

@Taylorminer 调用opencv库，对一组相机的内外参数进行标定。我用了2个相机，不是5个。

**Comment by Taylorminer** [comment]
> > 
> 
> @Taylorminer 调用opencv库，对一组相机的内外参数进行标定。我用了2个相机，不是5个。

谢谢您的回复，请问除了相机参数需要改变之外，在程序中还有其他地方需要进行修改吗？
